### PR TITLE
Up dev issue#10388

### LINKED ISSUE #10388: [type: new feature] support custom reject in MemorySafeLinkedBlockingQueue
- [x] I have searched the [issues](https://github.com/apache/dubbo/issues) of this repository and believe that this is not a duplicate.
- [x] I have searched the [release notes](https://github.com/apache/dubbo/releases) of this repository and believe that this is not a duplicate.

## Describe the feature

see ISSUE: https://github.com/apache/shenyu/issues/3763

Is there anyone willing to copy this PR:https://github.com/apache/shenyu/pull/3764 to Dubbo ?

### PR DESCRIPTION
**Pull request by TigerTurbo**
## What is the purpose of the change

fix #10388

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,7 +10,8 @@
 [![Twitter Follow](https://img.shields.io/twitter/follow/ApacheDubbo.svg?label=Follow&style=social&logoWidth=0)](https://twitter.com/intent/follow?screen_name=ApacheDubbo)
 [![Gitter](https://badges.gitter.im/alibaba/dubbo.svg)](https://gitter.im/alibaba/dubbo?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)
 
-Apache Dubbo is a high-performance, Java-based open-source RPC framework. Please visit the [official site](http://dubbo.apache.org) for the quick start guide and documentation, as well as the [wiki](https://github.com/apache/dubbo/wiki) for news, FAQ, and release notes.
+Apache Dubbo is a high-performance, Java-based open-source RPC framework. Please visit the [official site](http://dubbo.apache.org) for 
+the quick start guide and documentation, as well as the [wiki](https://github.com/apache/dubbo/wiki) for news, FAQ, and release notes.
```
Comment (AlbumenJ) [inline README.md:14]: why change this line

File: dubbo-common/src/main/java/org/apache/dubbo/common/concurrent/Rejector.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+     *
+     * <p>In the absence of other alternatives, the method may throw an unchecked
+     * {@link RejectException}, which will be propagated to the caller.
+     *
+     * @param e     the element requested to be added
+     * @param queue the queue attempting to add this element
+     *
+     * @throws RejectException if there is no more memory
+     */
+    void reject(E e, MemorySafeLinkedBlockingQueue<E> queue);
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/common/concurrent/Rejector.java:42]: `MemorySafeLinkedBlockingQueue` => `Queue` would be better?

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-common/src/main/java/org/apache/dubbo/common/concurrent/Rejector.java (hunk 2 above)
Comment (TigerTurbo) [reply dubbo-common/src/main/java/org/apache/dubbo/common/concurrent/Rejector.java:42]: if there is no more memory will be throw RejectException

File: dubbo-common/src/main/java/org/apache/dubbo/common/threadpool/MemorySafeLinkedBlockingQueue.java [hunk 3]
```diff
@@ -69,6 +78,15 @@ public int getMaxFreeMemory() {
...
     }
 
+    /**
+     * set the rejector.
+     *
+     * @param rejector the rejector
+     */
+    public void setRejector(final Rejector<E> rejector) {
+        this.rejector = rejector;
+    }
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/common/threadpool/MemorySafeLinkedBlockingQueue.java:88]: How can the user change this field?

File: dubbo-common/src/main/java/org/apache/dubbo/common/concurrent/Rejector.java (hunk 2 above)
Comment (AlbumenJ) [reply dubbo-common/src/main/java/org/apache/dubbo/common/concurrent/Rejector.java:42]: I mean it's better to change the signature of this function to `void reject(E e, Queue<E> queue)`. The `Rejector` here should not be coupled with actual queue types like `MemorySafeLinkedBlockingQueue`.

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

**Comment by TigerTurbo** [comment]
ok

File: dubbo-common/src/main/java/org/apache/dubbo/common/threadpool/MemorySafeLinkedBlockingQueue.java (hunk 3 above)
Comment (TigerTurbo) [reply dubbo-common/src/main/java/org/apache/dubbo/common/threadpool/MemorySafeLinkedBlockingQueue.java:88]: done

Comment (TigerTurbo) [reply dubbo-common/src/main/java/org/apache/dubbo/common/threadpool/MemorySafeLinkedBlockingQueue.java:88]: new出来这个队列，再set进去，如下图所示
image

**Pull request merged**